src/ecs/resources/rendering.rs

```diff
@@ -0,0 +1,110 @@
+use std::cell::UnsafeCell;
+use std::sync::Arc;
+
+use crossbeam::sync::MsQueue;
+use futures::{Async, Future, Poll};
+use renderer::{Error, Mesh, MeshBuilder};
+
+/// A factory future.
+pub struct FactoryFuture<A, E> {
+    inner: Option<Arc<FactoryFutureCell<A, E>>>,
+}
+
+impl<A: 'static, E: 'static> FactoryFuture<A, E> {
+    fn new() -> (Self, Arc<FactoryFutureCell<A, E>>) {
+        let inner = FactoryFutureCell { value: UnsafeCell::new(None) };
+        let inner = Arc::new(inner);
+
+        let cloned = inner.clone();
+        let inner = Some(inner);
+
+        (FactoryFuture { inner }, cloned)
+    }
+}
+
+impl<A, E> Future for FactoryFuture<A, E> {
+    type Item = A;
+    type Error = E;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match Arc::try_unwrap(self.inner.take().unwrap()) {
+            Ok(x) => {
+                // As soon as the worker thread finished executing the closure, the second
+                // reference gets dropped and we enter this branch.
+                // Thus, we are the only one with access to the cell.
+                unsafe {
+                    x.value
+                        .into_inner()
+                        .expect("Thread panicked")
+                        .map(Async::Ready)
+                }
+            }
+            Err(arc) => {
+                self.inner = Some(arc);
+
+                Ok(Async::NotReady)
+            }
+        }
+    }
+
+    fn wait(mut self) -> Result<Self::Item, Self::Error> {
+        use futures::Async;
+
+        loop {
+            match self.poll() {
+                Ok(Async::Ready(x)) => return Ok(x),
+                Ok(Async::NotReady) => {}
+                Err(x) => return Err(x),
+            }
+        }
+    }
+}
+
+struct FactoryFutureCell<A, E> {
+    value: UnsafeCell<Option<Result<A, E>>>,
+}
+
+impl<A, E> FactoryFutureCell<A, E> {
+    fn set(&self, res: Result<A, E>) {
```

Comment:
Such method itself is unsafe.
Common technique to have safe `set` method is to make special type that holds `Arc` and does not implement `Sync`, only `Send`